Add first draft of Airflow documentation (#128)
datamade/how-to: etl/airflow.md

preferred
solution](https://github.com/datamade/scrapers-us-municipal/blob/c537c27dbd949cf6b838612325f087b2e3860252/scripts/la-metro-crontask)
at DataMade!

However, cron contributed to or made it more difficult to solve some tough
problems for our scrapers, including [scheduling
typos](https://github.com/datamade/scrapers-us-municipal/pull/50),
[colliding](https://github.com/datamade/scrapers-us-municipal/issues/38)
[scrapes](https://github.com/datamade/scrapers-us-municipal/issues/3), and
[indefinitely hung
processes](https://github.com/datamade/scrapers-us-municipal/issues/55). Beyond
bugs and developer experience, our usage of cron did not give our clients any
insight into whether scrapes had run, and to what outcome.

What we wanted was a framework that would allow us to define, manage, and
monitor our ETL jobs, and also expose an interface for interested parties to
follow along. We investigated three candidates:
[Airflow](https://airflow.apache.org/index.html),
[Luigi](https://github.com/spotify/luigi), and
[Rundeck](https://www.rundeck.com/open-source). Rundeck seemed optimized more
for server administration than data workflows, so we mostly considered Airflow

> vs. Luigi.** **

Reviewer: what are these extra asterisks?

> vs. Luigi.** **

Reviewer: these asterisks don't seem to be doing anything here

> vs. Luigi.** **

Author: ```suggestion
vs. Luigi.
```

[This article](https://towardsdatascience.com/data-pipelines-luigi-airflow-everything-you-need-to-know-18dc741449b7)
provides an excellent comparison of Airflow and Luigi. The important

> [distinctions for us were as follows:

Reviewer: same with this square bracket w/o a buddy

> [distinctions for us were as follows:

Author: ```suggestion
distinctions for us were as follows:
```

1. Airflow supports scheduling tasks, while Luigi [purposefully omits this
functionality](https://luigi.readthedocs.io/en/stable/central_scheduler.html).
2. Luigi is a build tool. [Its units of
work](https://luigi.readthedocs.io/en/stable/tasks.html) are more Make-like in
that they are tightly coupled to output from their composite targets.
Conversely, Airflow operators run independently of one another, i.e., Airflow
feels more conceptually appropriate for scheduling tasks that don’t necessarily
create file-like artifacts, like scraping data.
3. Airflow has eclipsed Luigi in
popularity and maintenance resources, as measured by forks, stars, and watchers,
and commits, respectively.

Based on these impressions, and [an initial proof of
concept](https://github.com/datamade/scrapers-airflow), we selected Airflow as
our tool of choice for ETL management. Our first Airflow instance is the LA
Metro Dashboard, which has been open sourced
[here](https://github.com/datamade/la-metro-dashboard).

## Basic concepts

> Airflow is a framework that relies on several key concepts. For a tutorial that
> guides you through a summary of these concepts in Airflow’s words, go to
> [https://airflow.apache.org/docs/stable/tutorial.html#example-pipeline-definition](https://airflow.apache.org/docs/stable/tutorial.html#example-pipeline-definition).

Author: @fatima3558 Can you integrate these full URLs with the prose as hyperlinks, throughout this section? For example:

```suggestion
Airflow is a framework that relies on several key concepts. See [the Airflow
documentation](https://airflow.apache.org/docs/stable/tutorial.html#example-pipeline-definition) 
for a guided tour of these concepts in Airflow's own words.
```

Below are listed the most important Airflow concepts based on our work with
Airflow thus far.

> ### DAG (directed acyclic graph)

Author: @fatima3558 What do you think about reorganizing these concept subheds to follow the format:

### Concept

**Airflow's description**: Link

[Discussion]

#### Notes

- ...

A DAG is a way of visualizing the way Airflow works. It emphasizes that Airflow
moves through tasks in a particular order without automatically repeating those
tasks, and includes information about tasks’ dependencies on each other.

- Airflow’s description:
- [https://airflow.apache.org/docs/stable/concepts.html#dags](https://airflow.apache.org/docs/stable/concepts.html#dags)
- A DAG run is an instance of a DAG.

### Task

A task is a single function or command to run. At the bare minimum,
instantiating a task requires a `task_id`, `dag`, and the command argument for
the particular operator being used.

> - For example, the `bash_operator` has a `command` argument, whereas the
> - `python_operator` has a `python_callable` argument. A task run is an instance
> - of a task.

Author: Think this was a formatting goof on my part.

```suggestion
For example, the `bash_operator` has a `command` argument, whereas the
`python_operator` has a `python_callable` argument. A task run is an instance
of a task.
```

### Operator

Airflow is really powerful in part because it can run commands for any part of
your app that you can write in the most convenient language for that task [not

> sure about wording here]. For example, you can write a scraping script in Bash

Author: FWIW, I think this is great wording! Any thoughts, @jeancochrane?

and run it with the `bash_operator`, run a Python command using the
`python_operator`, and then run a database command using the `postgres_operator`
-- all in the same DAG!

There are also operators that handle control flow and other kinds of tasks. For
a full list, see
[https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html](https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html)

> 

Author: While Airflow supports many operators, we have up til now preferred the Docker operator for the reasons outlined in the Dependency Management section, below. Think it's worth stating that and adding a pointer to the later section, @fatima3558?

### Pipeline

> A configuration file for a DAG. There are five parts to a pipeline, and they

Author: @fatima3558 I think it would be useful to put a link up top to Airflow's example! https://airflow.apache.org/docs/stable/tutorial.html#example-pipeline-definition

the documentation for that specific operator for the most detailed information
on how to define a task. Here’s the list of operators:
[https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html](https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html)

    Keep in mind that there are also operators that are focused on control flow
    that you might need to use in defining tasks. A couple common ones you might
    use:

    - Branch Operator: Allows you to run different tasks at different times or
    under different conditions
    - Dummy Operator: If you want to avoid running a task if a certain condition
    is true, you can use the Dummy Operator to run a fake task in place of the
    true one.

5. **Task ordering and dependencies:** The final part of the pipeline is task
ordering. Here, set the order that the tasks should run in and how they depend
on each other. The example code given on the Airflow docs is concise and
helpful:
[https://airflow.apache.org/docs/stable/tutorial.html#setting-up-dependencies](https://airflow.apache.org/docs/stable/tutorial.html#setting-up-dependencies)

> ## Recommended default settings

Author: @jeancochrane Fatima indicated this section might be incomplete. Can you review?

Reviewer 2: Good catch on the connection string @fatima3558!

I just ran some diffs between the [default config for Airflow 1.10.10](https://raw.githubusercontent.com/apache/airflow/96697180d79bfc90f6964a8e99f9dd441789177c/airflow/config_templates/default_airflow.cfg) and our local config (`airflow.cfg`) and production config (`configs/airflow.production.cfg`) for https://github.com/datamade/la-metro-dashboard/. Here are some other changes that I found:

### Local `airflow.cfg`

```cfg
# Don't load example DAGs
load_examples = False

# Turn on authentication
authenticate = True

# Turn on role-based access control (Airflow's new model of authentication, with a separate UI)
rbac = True

# Don't catch up to old DAGs when restarting the process
catchup_by_default = False
```

We may not always want authentication and RBAC. I think that should probably be a project-by-project decision.

We also changed `executor` to `LocalExecutor` from `SequentialExecutor` but I don't actually think we should do this going forward because [`SequentialExecutor` is just `LocalExecutor` with parallelism set to `1`, which is probably good for local testing](https://airflow.apache.org/docs/stable/executor/local.html).

### Production `configs/airflow.production.cfg`

```cfg
# Store logs in /var/log intead of AIRFLOW_HOME
base_log_folder = /var/log/la-metro-dashboard/airflow
dag_processor_manager_log_location = /var/log/la-metro-dashboard/airflow/dag_processor_manager/dag_processor_manager.log

# Use a single node for task execution
executor = LocalExecutor

# Only two task instances should be able to run simultaneously
parallelism = 2

# The scheduler should only be able to run two task instances
dag_concurrency = 2

# Workers should only be able to run two task instances
worker_concurrency = 2

# Don't load example DAGs
load_examples = False

# Turn on authentication
authenticate = True

# Turn on role-based access control (Airflow's new model of authentication, with a separate UI)
rbac = True

# Don't catch up to old DAGs when restarting the process
catchup_by_default = False
```

Again, authentication and RBAC should be set project-by-project. The concurrency configs are super confusing and [the docs](https://airflow.apache.org/docs/stable/configurations-ref.html) don't really help so I think there's a lot of opportunity for better tuning in the future.

the code and precludes manually triggering DAGs, as the scheduling logic may not
yield the desired path.

We prefer to consolidate scheduling concerns and use only the schedule interval
to schedule DAG runs. We are [actively working on identifying a preferred
strategy](https://github.com/datamade/la-metro-dashboard/issues/59) for doing
this while minimizing repeat code and clever architecture.

## Deploying Airflow

In theory, Airflow should be easy to deploy on Heroku. There are a number of
guides for doing this online (like [this
one](https://medium.com/@damesavram/running-airflow-on-heroku-ed1d28f8013d)).
Most use Procfiles instead of the container stack, but our [docs for deploying
Django on Heroku using the container
stack](https://github.com/datamade/how-to/blob/master/heroku/deploy-a-django-app.md)
should be straightforward to adapt for a Flask app like Airflow.

However, our first production Airflow instance, the [LA Metro
Dashboard](https://github.com/datamade/la-metro-dashboard/), is deployed on AWS
EC2 via CodeDeploy using DataMade’s [legacy deployment
framework](https://github.com/datamade/deploy-a-site/). The primary reason we

> had to fall back to our legacy framework is because **we** **needed a stable IP

Author: ```suggestion
had to fall back to our legacy framework is because **we needed a stable IP
```

address that we could safelist to access an external Solr instance**. Our
original plan was to expose our Solr instance publicly and enable
[SSL](https://lucene.apache.org/solr/guide/7_6/enabling-ssl.html) and [basic
authentication](https://lucene.apache.org/solr/guide/8_1/basic-authentication-plugin.html)
so that an Airflow instance on Heroku could communicate with it, but we were
unable to get either feature to work properly before we ran out of budget for
the task.

Since the IP address of a Heroku app can change without notice, Heroku is a bad
choice for an Airflow deployment that requires a static IP address. However,
this is currently the only hurdle to deploying Airflow on Heroku that we know
of, and we may be able to get around it with more research.

## Tips and Tricks

Since Airflow is a packaged Flask app, debugging it can be challenging. The
following is an evolving list of tricks that we have identified for debugging
Airflow.

### Starting a Flask shell

The following command will start a [Flask
shell](https://flask.palletsprojects.com/en/1.1.x/shell/) in the Airflow app:

``` bash
# This command assumes your Airflow service in docker-compose.yml is

> named “airflow”

Author: ```suggestion
# named “airflow”
```

docker-compose run --rm -e FLASK_APP=airflow.www.app airflow flask shell
```

### Preventing worker refresh when debugging

Airflow needs to periodically refresh its worker process in order to register
changes to DAGS, and by default it does this every 30 seconds. This can be
frustrating during debugging because refreshing the worker process will end a
running process and kick you out of a debugger that you’ve dropped into that
process.

To prevent reloading when debugging, you can temporarily set the

> [`worker_refresh_interval`](https://github.com/apache/airflow/blob/11eb649d4acdbd3582fb0a77b5f5af3b75e2262c/airflow/config_templates/default_airflow.cfg#L430-L431)

Author: @jeancochrane I think it would be a good idea to point to this, and the reloading plugin code, in the section on default settings.
